services/docker/docker-size.service.js

```diff
@@ -73,30 +84,77 @@ export default class DockerSize extends BaseJsonService {
     })
   }
 
-  transform({ tag, sort, data }) {
+  transform({ tag, sort, data, arch }) {
+    let sizeFromSpecifiedArchitecture
+
     if (!tag && sort === 'date') {
       if (data.count === 0) {
         throw new NotFound({ prettyMessage: 'repository not found' })
       } else {
-        return { size: data.results[0].full_size }
+        const latestEntry = data.results[0]
+
+        // If no tag is specified, and sorting is by date, check if any of the returned images has an architecture matching the arch parameter supplied by the user.
+        // If yes, return the size of the image with this arch.
+        // If not, return the value of the `full_size` in the latestEntry from the response.
+        // For details see: https://github.com/badges/shields/issues/8238
+
+        Object.values(latestEntry.images).forEach(img => {
+          if (img.architecture === arch) {
+            sizeFromSpecifiedArchitecture = img.size
+          }
+        })
```

Comment:
There's 3 places in this file where we're doing basically the same thing with slightly different variables. You could pull this out to a helper function and then call it in 3 places to DRY up:


```js
function getImageForArch(images, arch)
  Object.values(images).forEach(img => {
    if (img.architecture === arch) {
      return img
    }
  })
  // TODO: if we get here throw a `NotFound()`
}
```

Also notice the early return as soon as we find the result. We don't need to keep iterating the array after we've found what we want. We could have also done this with a `break;`.

Reply:
Very good point, thanks!

Updated to look like this:

```js
function getImageForArch(images, arch) {
  const imgWithArch = Object.values(images).find(
    img => img.architecture === arch
  )

  if (!imgWithArch) {
    throw new NotFound({ prettyMessage: 'architecture not found' })
  }
  return imgWithArch.size
}

```

Since the `find()` method returns the first element in the provided array that satisfies the provided testing function, no unnecessary iteration would be done.